### PR TITLE
[Improve] Speed up data preprocessor.

### PR DESCRIPTION
**Pull request by mzr1996**
## Motivation

The original data preprocessor consumes much time because it needs to cast the label tensors of samples one by one.

## Modification

1. In `ClsDataPreprocessoor`, only cast `gt_label` of data samples to cuda instead of the whole data sample. And before casting, stack/concat the tensors.
2. Refactoring the batch augmentations class like `Mixup`. Now it directly processes the batch inputs and batch scores (one-hot format labels), and won't accept normal labels or data samples.
3. Add ClsDataSample serialization override functions. During serialization in `ForkingPicker`, convert all tensors to NumPy array and convert them back during deserialization. This is to decrease the consumption of file descriptors in the dataloader.

Here is the speed comparison in MobileNetV2 (batch size 64)
|    | inference time | FPS |
|:--:|:--:|:--:|
| Original | 19.5 ms | 3289  |
| New |  14.1 ms | 4531|

## BC-breaking (Optional)

The `Mixup`, `CutMix`, and `ResizeMix` classes won't accept the `num_classes` argument, and it becomes the argument of `ClsDataPreprocessor`.

In the config files, the changes are as below:
```python
# --------- Original config --------
model = dict(
    ...
    train_cfg=dict(augments=[
        dict(type='Mixup', alpha=0.2, num_classes=1000),
        dict(type='CutMix', alpha=1.0, num_classes=1000)
    ]),
)

data_preprocessor = dict(
    mean=[123.675, 116.28, 103.53],
    std=[58.395, 57.12, 57.375],
    to_rgb=True,
)

# -------- New config --------
model = dict(
    ...
    train_cfg=dict(augments=[
        dict(type='Mixup', alpha=0.2,
        dict(type='CutMix', alpha=1.0)
    ]),
)

data_preprocessor = dict(
    num_classes=1000,
    mean=[123.675, 116.28, 103.53],
    std=[58.395, 57.12, 57.375],
    to_rgb=True,
)
```

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: configs/_base_/datasets/imagenet21k_bs128.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # dataset settings
 dataset_type = 'ImageNet21k'
 data_preprocessor = dict(
+    num_classes=1000,
```
Comment (tonysy) [inline configs/_base_/datasets/imagenet21k_bs128.py:4]: imagenet 21k should have ~21k classes?

File: configs/_base_/datasets/imagenet_bs64_mixer_224.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
```
Comment (tonysy) [inline configs/_base_/datasets/imagenet_bs64_mixer_224.py:3]: This line can be removed to keep consistent with other config file

File: mmcls/models/utils/data_preprocessor.py [hunk 3]
```diff
@@ -135,12 +150,36 @@ def forward(self, data: dict, training: bool = False) -> dict:
...
-        # ----- Batch Aug ----
-        if training and self.batch_augments is not None:
-            data_samples = data['data_samples']
-            inputs, data_samples = self.batch_augments(inputs, data_samples)
-            data['data_samples'] = data_samples
-
-        data['inputs'] = inputs
-
-        return data
+        data_samples = data.get('data_samples', None)
```
Comment (tonysy) [inline mmcls/models/utils/data_preprocessor.py:153]: To help the user for understanding our codebase, maybe we need to update the document by adding some description of the `data_samples` used in MMClassification(with text or link to MMEngine related document), this could be added into the document TODO list.

Comment (tonysy) [reply mmcls/models/utils/data_preprocessor.py:153]: We can create an issue and add it to Oct. TODO list

**Review by tonysy — APPROVED** [review]
LGTM
